// @lit-labs/motion/development/animate-controller.js
export const controllerMap = new WeakMap();
/**
 * AnimateController can be used to provide default configuration options to all
 * `animate()` directives in a given component.
 *
 * It also provides control over all the `animate()` animations within a given
 * component, for example, `this.animateController.pause()` pauses all animations.
 */
export class AnimateController {
    constructor(host, options) {
        this.startPaused = false;
        this.disabled = false;
        /**
         * Set of active `animate()` directives in the host component
         */
        this.clients = new Set();
        this.pendingComplete = false;
        this.host = host;
        this.defaultOptions = options.defaultOptions || {};
        this.startPaused = !!options.startPaused;
        this.disabled = !!options.disabled;
        this.onComplete = options.onComplete;
        controllerMap.set(this.host, this);
    }
    async add(animate) {
        var _a, _b;
        this.clients.add(animate);
        if (this.startPaused) {
            (_a = animate.webAnimation) === null || _a === void 0 ? void 0 : _a.pause();
        }
        this.pendingComplete = true;
        await animate.finished;
        if (this.pendingComplete && !this.isAnimating) {
            this.pendingComplete = false;
            (_b = this.onComplete) === null || _b === void 0 ? void 0 : _b.call(this);
        }
    }
    remove(animate) {
        this.clients.delete(animate);
    }
    /**
     * Pauses all animations running in the host component.
     */
    pause() {
        this.clients.forEach((f) => { var _a; return (_a = f.webAnimation) === null || _a === void 0 ? void 0 : _a.pause(); });
    }
    /**
     * Plays all active animations in the host component.
     */
    play() {
        this.clients.forEach((f) => { var _a; return (_a = f.webAnimation) === null || _a === void 0 ? void 0 : _a.play(); });
    }
    cancel() {
        this.clients.forEach((f) => { var _a; return (_a = f.webAnimation) === null || _a === void 0 ? void 0 : _a.cancel(); });
        this.clients.clear();
    }
    finish() {
        this.clients.forEach((f) => { var _a; return (_a = f.webAnimation) === null || _a === void 0 ? void 0 : _a.finish(); });
        this.clients.clear();
    }
    /**
     * Toggles the play/pause state of all active `animate()` animations in the host component.
     */
    togglePlay() {
        if (this.isPlaying) {
            this.pause();
        }
        else {
            this.play();
        }
    }
    /**
     * Returns true if the host component has any active `animate()` animations.
     */
    get isAnimating() {
        return this.clients.size > 0;
    }
    /**
     * Returns true if the host component has any playing `animate()` animations.
     */
    get isPlaying() {
        return Array.from(this.clients).some((a) => { var _a; return ((_a = a.webAnimation) === null || _a === void 0 ? void 0 : _a.playState) === 'running'; });
    }
    async finished() {
        await Promise.all(Array.from(this.clients).map((f) => f.finished));
    }
}
