// @storybook/builder-vite/dist/plugins/mdx-plugin.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mdxPlugin = void 0;
const isStorybookMdx = (id) => id.endsWith('stories.mdx') || id.endsWith('story.mdx');
function injectRenderer(code, mdx2) {
    if (mdx2) {
        return `
           import React from 'react';
           ${code}
           `;
    }
    return `
        /* @jsx mdx */
        import React from 'react';
        import { mdx } from '@mdx-js/react';
        ${code}
        `;
}
/**
 * Storybook uses two different loaders when dealing with MDX:
 *
 * - *stories.mdx and *story.mdx are compiled with the CSF compiler
 * - *.mdx are compiled with the MDX compiler directly
 *
 * @see https://github.com/storybookjs/storybook/blob/next/addons/docs/docs/recipes.md#csf-stories-with-arbitrary-mdx
 */
function mdxPlugin(options) {
    const { features } = options;
    let reactRefresh;
    return {
        name: 'storybook-vite-mdx-plugin',
        enforce: 'pre',
        configResolved({ plugins }) {
            // @vitejs/plugin-react-refresh has been upgraded to @vitejs/plugin-react,
            // and the name of the plugin performing `transform` has been changed from 'react-refresh' to 'vite:react-babel',
            // to be compatible, we need to look for both plugin name.
            // We should also look for the other plugins names exported from @vitejs/plugin-react in case there are some internal refactors.
            const reactRefreshPlugins = plugins.filter((p) => p.name === 'react-refresh' ||
                p.name === 'vite:react-babel' ||
                p.name === 'vite:react-refresh' ||
                p.name === 'vite:react-jsx');
            reactRefresh = reactRefreshPlugins.find((p) => p.transform);
        },
        async transform(src, id, options) {
            if (id.match(/\.mdx?$/)) {
                // @ts-ignore
                const { compile } = (features === null || features === void 0 ? void 0 : features.previewMdx2)
                    ? await Promise.resolve().then(() => __importStar(require('@storybook/mdx2-csf')))
                    : await Promise.resolve().then(() => __importStar(require('@storybook/mdx1-csf')));
                const mdxCode = String(await compile(src, { skipCsf: !isStorybookMdx(id) }));
                const modifiedCode = injectRenderer(mdxCode, Boolean(features === null || features === void 0 ? void 0 : features.previewMdx2));
                const result = await (reactRefresh === null || reactRefresh === void 0 ? void 0 : reactRefresh.transform.call(this, modifiedCode, `${id}.jsx`, options));
                if (!result)
                    return modifiedCode;
                if (typeof result === 'string')
                    return result;
                const { code, map: resultMap } = result;
                return { code, map: !resultMap || typeof resultMap === 'string' ? resultMap : { ...resultMap, sources: [id] } };
            }
        },
    };
}
exports.mdxPlugin = mdxPlugin;
