// @web/test-runner/dist/reporter/summaryReporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.summaryReporter = void 0;
const reportTestsErrors_js_1 = require("./reportTestsErrors.js");
const reportTestFileErrors_js_1 = require("./reportTestFileErrors.js");
const TestRunnerLogger_js_1 = require("../logger/TestRunnerLogger.js");
const color = ([x, y]) => (z) => `\x1b[${x}m${z}\x1b[${y}m${reset}`;
const reset = `\x1b[0m\x1b[0m`;
const green = color([32, 89]);
const red = color([31, 89]);
/** Test reporter that summarizes all test for a given run */
function summaryReporter(opts) {
    const { flatten = false } = opts !== null && opts !== void 0 ? opts : {};
    let args;
    let favoriteBrowser;
    const logger = new TestRunnerLogger_js_1.TestRunnerLogger();
    function log(name, passed, prefix = '  ') {
        const sign = passed ? green('âœ“') : red('ð„‚');
        logger.log(flatten ? `${sign} ${prefix} ${name}` : `${prefix}  ${sign} ${name}`);
    }
    function logResults(results, prefix) {
        var _a, _b;
        for (const result of (_a = results === null || results === void 0 ? void 0 : results.tests) !== null && _a !== void 0 ? _a : []) {
            log(result.name, result.passed, prefix);
        }
        for (const suite of (_b = results === null || results === void 0 ? void 0 : results.suites) !== null && _b !== void 0 ? _b : []) {
            logSuite(suite, prefix);
        }
    }
    function logSuite(suite, parent) {
        const pref = flatten ? `${parent ? `${parent}` : ''} ${suite.name}` : `${parent !== null && parent !== void 0 ? parent : ''} `;
        if (!flatten) {
            logger.log(`${pref}${suite.name}`);
        }
        logResults(suite, pref);
    }
    return {
        start(_args) {
            var _a;
            args = _args;
            favoriteBrowser = (_a = args.browserNames.find(name => {
                const n = name.toLowerCase();
                return n.includes('chrome') || n.includes('chromium') || n.includes('firefox');
            })) !== null && _a !== void 0 ? _a : args.browserNames[0];
        },
        reportTestFileResults({ sessionsForTestFile: sessions }) {
            for (const session of sessions) {
                logResults(session.testResults);
                logger.log('');
            }
        },
        onTestRunFinished({ sessions }) {
            const failedSessions = sessions.filter(s => !s.passed);
            if (failedSessions.length > 0) {
                logger.log('\n\nErrors Reported in Tests:\n\n');
                reportTestsErrors_js_1.reportTestsErrors(logger, args.browserNames, favoriteBrowser, failedSessions);
                reportTestFileErrors_js_1.reportTestFileErrors(logger, args.browserNames, favoriteBrowser, failedSessions, true);
            }
        },
    };
}
exports.summaryReporter = summaryReporter;
