// @storybook/web-components/dist/cjs/client/index.js
"use strict";

require("core-js/modules/es.object.to-string.js");

require("core-js/modules/web.dom-collections.for-each.js");

require("core-js/modules/es.object.keys.js");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  storiesOf: true,
  setAddon: true,
  addDecorator: true,
  addParameters: true,
  configure: true,
  getStorybook: true,
  forceReRender: true,
  raw: true,
  getCustomElements: true,
  setCustomElements: true,
  setCustomElementsManifest: true,
  isValidComponent: true,
  isValidMetaData: true
};
Object.defineProperty(exports, "addDecorator", {
  enumerable: true,
  get: function get() {
    return _preview.addDecorator;
  }
});
Object.defineProperty(exports, "addParameters", {
  enumerable: true,
  get: function get() {
    return _preview.addParameters;
  }
});
Object.defineProperty(exports, "configure", {
  enumerable: true,
  get: function get() {
    return _preview.configure;
  }
});
Object.defineProperty(exports, "forceReRender", {
  enumerable: true,
  get: function get() {
    return _preview.forceReRender;
  }
});
Object.defineProperty(exports, "getCustomElements", {
  enumerable: true,
  get: function get() {
    return _customElements.getCustomElements;
  }
});
Object.defineProperty(exports, "getStorybook", {
  enumerable: true,
  get: function get() {
    return _preview.getStorybook;
  }
});
Object.defineProperty(exports, "isValidComponent", {
  enumerable: true,
  get: function get() {
    return _customElements.isValidComponent;
  }
});
Object.defineProperty(exports, "isValidMetaData", {
  enumerable: true,
  get: function get() {
    return _customElements.isValidMetaData;
  }
});
Object.defineProperty(exports, "raw", {
  enumerable: true,
  get: function get() {
    return _preview.raw;
  }
});
Object.defineProperty(exports, "setAddon", {
  enumerable: true,
  get: function get() {
    return _preview.setAddon;
  }
});
Object.defineProperty(exports, "setCustomElements", {
  enumerable: true,
  get: function get() {
    return _customElements.setCustomElements;
  }
});
Object.defineProperty(exports, "setCustomElementsManifest", {
  enumerable: true,
  get: function get() {
    return _customElements.setCustomElementsManifest;
  }
});
Object.defineProperty(exports, "storiesOf", {
  enumerable: true,
  get: function get() {
    return _preview.storiesOf;
  }
});

var _global = _interopRequireDefault(require("global"));

var _preview = require("./preview");

var _customElements = require("./customElements");

var _types = require("./preview/types-6-0");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _types[key];
    }
  });
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var window = _global.default.window,
    EventSource = _global.default.EventSource;

// TODO: disable HMR and do full page loads because of customElements.define
if (module && module.hot && module.hot.decline) {
  module.hot.decline(); // forcing full reloads for customElements as elements can only be defined once per page

  var hmr = new EventSource('__webpack_hmr');
  hmr.addEventListener('message', function fullPageReload(event) {
    try {
      // Only care for built events.  Heartbeats are not parsable so we ignore those
      var _JSON$parse = JSON.parse(event.data),
          action = _JSON$parse.action;

      if (action === 'built') {
        window.location.reload();
      }
    } catch (error) {// Most part we only get here from the data in the server-sent event not being parsable which is ok
    }
  });
}
