// @contentauth/config/eslint/default.cjs
module.exports = {
  root: true,
  parser: '@typescript-eslint/parser',
  ignorePatterns: [
    '**/*.d.ts',
    '**/*.spec.ts',
    '**/dist/*',
    '**/fixtures/*',
    '**/*.stories.ts',
  ],
  plugins: ['@typescript-eslint'],
  env: {
    es6: true,
    node: true,
  },
  extends: [
    'airbnb-typescript/base',
    'plugin:import/recommended',
    'plugin:import/typescript',
    'prettier',
  ],
  rules: {
    '@typescript-eslint/no-unused-vars': 'warn',
    '@typescript-eslint/no-use-before-define': 'off',
    'import/no-named-as-default': 'off',
    'class-methods-use-this': 'off',
    'function-paren-newline': 'off',
    'global-require': 'off',
    'implicit-arrow-linebreak': 'off',
    'import/extensions': 'off',
    'import/no-extraneous-dependencies': ['warn', { devDependencies: true }],
    'import/prefer-default-export': 'off',
    'no-confusing-arrow': 'off',
    'no-param-reassign': 'off',
    'no-underscore-dangle': 'off',
    'object-curly-newline': 'off',
    'operator-linebreak': 'off',
    'react/jsx-one-expression-per-line': 'off',
    'react/jsx-props-no-spreading': 'off',
  },
};
