// c2pa/rollup.config.js
/**
 * Copyright 2021 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it.
 */

import integrity from '@contentauth/toolkit/pkg/integrity.json';
import commonjs from '@rollup/plugin-commonjs';
import json from '@rollup/plugin-json';
import nodeResolve from '@rollup/plugin-node-resolve';
import replace from '@rollup/plugin-replace';
import strip from '@rollup/plugin-strip';
import typescript from '@rollup/plugin-typescript';
import url from '@rollup/plugin-url';
import { wasm } from '@rollup/plugin-wasm';
import merge from 'lodash/merge';
import { resolve } from 'path';
import copy from 'rollup-plugin-copy';
import { terser } from 'rollup-plugin-terser';

const banner = `
/*!*************************************************************************
 * Copyright 2021 Adobe
 * All Rights Reserved.
 *
 * NOTICE: Adobe permits you to use, modify, and distribute this file in
 * accordance with the terms of the Adobe license agreement accompanying
 * it. 
 **************************************************************************/
`;

const outputDir = resolve(__dirname, './dist');
const developmentMode = process.env.ROLLUP_WATCH === 'true';
const plugins = [
  copy({
    targets: [
      {
        src: [
          // Underneath copy plugin the `globby` package is used, which expects
          // paths to have`/` only, so the windows path separators generated by
          // path.resolve function cause it to break.
          resolve(
            __dirname,
            './node_modules/@contentauth/toolkit/pkg/**/*.wasm',
          ).replace(/\\/g, '/'),
        ],
        dest: resolve(__dirname, './dist/assets/wasm'),
      },
    ],
    copyOnce: false,
    flatten: true,
  }),
  wasm({
    maxFileSize: 1024000,
  }),
  nodeResolve({
    browser: true,
  }),
  commonjs(),
  typescript(),
  json(),
  replace({
    preventAssignment: true,
    values: {
      TOOLKIT_INTEGRITY: JSON.stringify(integrity),
    },
  }),
  url(),
];

const files = [
  {
    input: {
      c2pa: 'index.ts',
    },
    output: {
      format: 'es',
    },
  },
  {
    input: {
      'c2pa.worker': 'worker.ts',
    },
    output: {
      format: 'umd',
    },
  },
];

export default files.reduce((acc, config) => {
  const suffix = config.output.format === 'es' ? 'esm.' : '';
  const baseConfig = merge({}, config, {
    output: {
      entryFileNames: `[name].${suffix}js`,
      dir: outputDir,
      banner,
    },
  });
  acc.push(
    merge({}, baseConfig, {
      output: {
        sourcemap: 'inline',
      },
      plugins,
    }),
  );
  if (!developmentMode) {
    acc.push(
      merge({}, baseConfig, {
        output: {
          entryFileNames: `[name].${suffix}min.js`,
        },
        plugins: [
          ...plugins,
          strip({
            functions: ['dbg'],
          }),
          terser(),
        ],
      }),
    );
  }
  return acc;
}, []);
